File: 1647-minimum-deletions-to-make-character-frequencies-unique/1647-minimum-deletions-to-make-character-frequencies-unique.py
from collections import defaultdict
class Solution:
    def minDeletions(self, s: str) -> int:
        freq = defaultdict(int)  # frequency of each letter
        for c in s:
            freq[c] += 1
        count = defaultdict(int)  # number of letters per frequence
        for f in freq.values():
            count[f] += 1

        double_frequencies = True
        number_deletions = 0
        while double_frequencies:
            double_frequencies = False
            for f in list(count.keys()):
                if (f != 0) and (count[f] >= 2):
                    double_frequencies = True
                    number_deletions += (count[f] - 1)
                    count[f-1] += (count[f] - 1)
                    count[f] = 1
        return number_deletions

File: 1647-minimum-deletions-to-make-character-frequencies-unique/1647-minimum-deletions-to-make-character-frequencies-unique.cpp
class Solution {
public:
    int minDeletions(string s) {
        std::unordered_map<char, int> freq;  // frequency of each letter
        char c;
        for (int i=0; i<s.length(); i++) {
            c = s[i];
            if (freq.find(c) == freq.end()) {
                freq.insert({c, 0});
            }
            freq[c] += 1;
        }
        std::unordered_map<int, int> count;  // number of letters per frequence
        
        int f;
        for (auto& it: freq) {
            f = it.second;
            if (count.find(f) == count.end()) {
                count.insert({f, 0});
            }
            count[f] += 1;
        }

        bool double_frequencies = true;
        int number_deletions = 0;
        while (double_frequencies) {
            double_frequencies = false;
            for (auto& it: count) {
                f = it.first;
                if ((f != 0) and (count[f] >= 2)) {
                    double_frequencies = true;
                    number_deletions += (count[f] - 1);
                    count[f-1] += (count[f] - 1);
                    count[f] = 1;
                }
            }
        }
        return number_deletions;
    }
};